### PR TITLE
nixos/osk-sdl: init

### PR DESCRIPTION
**Pull request by tomfitzhenry**
###### Description of changes

This PR introduces:
* a module `boot.initrd.osk-sdl`, which provides an on-screen keyboard to unlock LUKS volumes during initrd, using [osk-sdl](https://gitlab.com/postmarketOS/osk-sdl) acting as a [systemd pasword agent](https://systemd.io/PASSWORD_AGENTS/). This allows osk-sdl to act as one password agent amongst many (TPM, FIDO, console, etc.)
* the osk-sdl package itself
* a test covering osk-sdl: `systemd-initrd-luks-osk-sdl`. See below for screenshots.

osk-sdl is already packaged on [Debian](https://tracker.debian.org/pkg/osk-sdl), [PostmarketOS](https://gitlab.com/postmarketOS/pmaports/-/tree/master/main/osk-sdl), and [more](https://repology.org/project/osk-sdl/versions).

Limitations:

* Currently only supported on `boot.initrd.systemd`, the systemd-based initrd. Additional support can be added later.
* The test is failing on aarch64-linux, as with all other aarch64-linux tests that use systemd-boot: https://github.com/NixOS/nixpkgs/issues/199843.

## Tests performed

Working on:

* a x86_64 physical machine
* a Pinephone
* QEMU x86_64.

Screenshots below.

<details>
  <summary>QEMU x86_64</summary>

prompt

pw

</details>

<details>
  <summary>Pinephone booting</summary>

pinephone

</details>

<details>
  <summary>A physical x86_64 machine</summary>

x86_64_desktop_prompt
x86_64_desktop_pw

</details>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/osk-sdl.nix [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          unitConfig.DefaultDependencies = false;
+          after = [
+            "systemd-vconsole-setup.service"
+          ];
+          before = [
+            "shutdown.target"
+          ];
+          script = ''
+            mkdir -p /run/opengl-driver/
+            cp -r ${pkgs.mesa.drivers}/* /run/opengl-driver/
```
[the remark this answers is not in the document]
Comment (zhaofengli) [reply nixos/modules/system/boot/osk-sdl.nix:59]: Maybe we should use `config.hardware.opengl.package` or [the `buildEnv` results](https://github.com/NixOS/nixpkgs/blob/9f1f990aee8ca8bc847bf41feabe0d91d8a59dc4/nixos/modules/hardware/opengl.nix#L13-L21) that combine both `package` and `extraPackages` but that would need to be exposed first.

**Comment by zhaofengli** [comment]
Cc @ElvishJerricco

**Comment by samueldr** [comment]
I was asked to chime-in on this subject:

Considering the reliance on HW-accelerated graphics being configured and working during such a critical step, this would also need to ensure there is robustness and will still work if HW-acceleration is broken in any way.

When evaluating options for Mobile NixOS, `osk-sdl` was rejected since it relies on SDL2, which can't work without HW acceleration. This could leave users in a tough situation to resolve.

File: nixos/modules/system/boot/osk-sdl.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+            export SDL_VIDEODRIVER=kmsdrm
+            for file in `ls /run/systemd/ask-password/*`; do
+              socket="$(cat "$file" | ${pkgs.gnugrep}/bin/grep "Socket=" | cut -d= -f2)"
+              ${pkgs.osk-sdl}/bin/osk-sdl -c ${pkgs.osk-sdl}/etc/osk.conf -d x -n x -k -s | ${pkgs.systemd}/lib/systemd/systemd-reply-password 1 "$socket"
+            done
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/osk-sdl.nix:65]: Looking at https://github.com/systemd/systemd/blob/main/src/tty-ask-password-agent/tty-ask-password-agent.c, this appears to be a very incomplete way to respond to password requests.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/osk-sdl.nix:65]: Also, the way this script is written, if a new password request is queued after the loop has begun, it won't be answered at all by this unit, I think. Unless the below `path` unit will restart this if the directory continues to be non-empty? I don't think it does though.

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]
This appears to only add support for qemu's graphics, and so wouldn't work on almost any hardware without further changes to include graphics drivers (and firmware) in stage 1.

Also, I agree with @samueldr. I think it is a big ask to require functional hardware accelerated graphics in stage 1. We have enough trouble with that in stage 2 as it is. I think it would be extremely difficult to make this reliably usable on a wide variety of hardware.

File: nixos/modules/system/boot/osk-sdl.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            "systemd-vconsole-setup.service"
+          ];
+          before = [
+            "shutdown.target"
+          ];
+          script = ''
+            mkdir -p /run/opengl-driver/
+            cp -r ${pkgs.mesa.drivers}/* /run/opengl-driver/
+
+            export SDL_VIDEODRIVER=kmsdrm
```
[the remark this answers is not in the document]
Comment (samueldr) [reply nixos/modules/system/boot/osk-sdl.nix:61]: https://github.com/libsdl-org/SDL/blob/15a9890919448cb463ab27f0e69be950be11ea87/include/SDL_hints.h#L2308-L2329

**Review by jokogr — COMMENTED** [review]
I have been also looking into this recently. My use case is a Surface Go 2 tablet, which is lacking the UEFI driver of an on-screen keyboard and therefore I can't decrypt the tablet's disk without connecting a keyboard at boot time.

@tomfitzhenry you've reached further from where I got, I was blocked on copying the Mesa drivers to the initrd image. I am still concerned about the initrd size, my tablet's boot partition is a small one.

Another project I had in mind for implementing this (on-screen keyboard for LUKS decrypting on stage 1) is unl0kr: https://gitlab.com/cherrypicker/unl0kr

* This is going to replace osk-sdl in postmarketOS, so maybe it's a better candidate for maintaining it on NixOS.
* It focuses just on inputting passwords, so the cryptsetup dependency is gone.
* It also uses LVGL instead of SDL, so it should lead to smaller initrd sizes. Mesa drivers are not even needed.
* My current struggle is its libinput dependency. I have it already built and packaged inside stage 1, but it appears that I can't use any input device (keyboard, pointer or multitouch) to control it. On normal NixOS (stage 2) it works as intended, though. Maybe I'm just missing a kernel module? :man_shrugging:

**Comment by samueldr** [comment]
@jokogr, as an alternative, though not integrated with either the classic or the systemd NixOS stage-1, the Mobile NixOS stage-1 is a drop-in replacement now. It supports unlocking LUKS volumes.

And from experience with different devices, yes you're likely missing the module(s) required for the touchscreen if LVGL can't use it. That's at least what happens with my libinput driver for LVGL.

File: nixos/modules/system/boot/osk-sdl.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (ElvishJerricco) [reply nixos/modules/system/boot/osk-sdl.nix:54]: Ah, my bad, `/run` will get overmounted at runtime.

**Comment by jokogr** [comment]
> > I'll take a look at unl0kr. Maybe it works well enough.
> 
> I'm close to getting that working. I'll close this for now, and re-open if needed.

Let me know if you need any beta testers :)

**Pull request closed without merging**